Jetty 12.1 Regression: Deferred Authentication provides Callback.NOT_CALLED leading to errors

**Jetty version(s)**
12.1.x

**Jetty Environment**
ee10

**HTTP version**
any

**Java version/vendor** `(use: java -version)`
25.0.3 Amazon Corretto

**OS type/version**
any

**Description**

Jetty 12.0 provided `Callback.NOOP` to `org.eclipse.jetty.security.Authenticator#validateRequest`. Any authentication could write a response. Deferred authentication ignored these responses, because the callback was `Callback.NOOP`.

Jetty 12.1 changed the callback to `Callback.NOT_CALLED`, resulting in an exception being thrown if a response is generated within the authenticator.

This is caused by:

 - https://github.com/jetty/jetty.project/issues/13012
 - https://github.com/jetty/jetty.project/commit/caa8fa1b9129085ac24e32fc4d6c923850c9ec91

The regression needs a few preconditions:

 - `org.eclipse.jetty.security.authentication.LoginAuthenticator` to have a deferred authentication
 - The authenticator needs to send a response itself in some situations.
 - Something must trigger the deferred authentication to become undeferred, in my case that's the framework calling `getAuthType()`:

```
	at com.example.auth.SimpleAuthenticator.validateRequest(SimpleAuthenticator.java:36)
	at org.eclipse.jetty.security.internal.DeferredAuthenticationState.authenticate(DeferredAuthenticationState.java:52)
	at org.eclipse.jetty.ee10.servlet.ServletApiRequest.getUndeferredAuthenticationState(ServletApiRequest.java:280)
	at org.eclipse.jetty.ee10.servlet.ServletApiRequest.getAuthType(ServletApiRequest.java:399)
	at jakarta.servlet.http.HttpServletRequestWrapper.getAuthType(HttpServletRequestWrapper.java:59)
	at org.stripesframework.web.controller.FlashRequest.<init>(FlashRequest.java:111)
```

**How to reproduce?**

I've created a full reproduction repository at https://github.com/kelunik/stripes-jetty-12.1-auth-bug. It's a very small app based on the [Stripes framework](https://github.com/Chrono24/stripes).

It can be started using the `Main` class:

<img width="253" height="102" alt="Image" src="https://github.com/user-attachments/assets/08946c06-92a3-4a03-b0b6-14ac0608709f" />

- Visit http://localhost:8080/example.action
- The first link works as expected, the second link "skips" the authentication within the authenticator and prints the following errors in the log:

```
[...] WARN org.eclipse.jetty.ee10.servlet.ServletChannel - /example.action
java.lang.IllegalStateException: java.lang.IllegalStateException: Committed
	at org.eclipse.jetty.util.Callback$2.failed(Callback.java:71)
	at org.eclipse.jetty.server.Response.writeError(Response.java:656)
	at org.eclipse.jetty.server.Response.writeError(Response.java:607)
	at com.example.auth.SimpleAuthenticator.validateRequest(SimpleAuthenticator.java:36)
	at org.eclipse.jetty.security.internal.DeferredAuthenticationState.authenticate(DeferredAuthenticationState.java:52)
	at org.eclipse.jetty.ee10.servlet.ServletApiRequest.getUndeferredAuthenticationState(ServletApiRequest.java:280)
	at org.eclipse.jetty.ee10.servlet.ServletApiRequest.getAuthType(ServletApiRequest.java:399)
	at jakarta.servlet.http.HttpServletRequestWrapper.getAuthType(HttpServletRequestWrapper.java:59)
	at org.stripesframework.web.controller.FlashRequest.<init>(FlashRequest.java:111)
	at org.stripesframework.web.controller.FlashRequest.replaceRequest(FlashRequest.java:62)
	at org.stripesframework.web.controller.FlashScope.completeRequest(FlashScope.java:349)
	at org.stripesframework.web.controller.StripesFilter.flashOutbound(StripesFilter.java:294)
	at org.stripesframework.web.controller.StripesFilter.doFilter(StripesFilter.java:233)
	at org.eclipse.jetty.ee10.servlet.FilterHolder.doFilter(FilterHolder.java:208)
	at org.eclipse.jetty.ee10.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1594)
	at org.eclipse.jetty.ee10.servlet.ServletHandler$MappedServlet.handle(ServletHandler.java:1556)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.dispatch(ServletChannel.java:871)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.handle(ServletChannel.java:449)
	at org.eclipse.jetty.ee10.servlet.ServletHandler.handle(ServletHandler.java:469)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:546)
	at org.eclipse.jetty.ee10.servlet.SessionHandler.handle(SessionHandler.java:719)
	at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1226)
	at org.eclipse.jetty.server.Server.handle(Server.java:197)
	at org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:788)
	at org.eclipse.jetty.server.internal.HttpConnection.onFillable(HttpConnection.java:420)
	at org.eclipse.jetty.server.internal.HttpConnection$FillableCallback.succeeded(HttpConnection.java:1809)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:54)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:492)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.epcRunTask(AdaptiveExecutionStrategy.java:428)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:401)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:255)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run(AdaptiveExecutionStrategy.java:204)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:317)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:1009)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1240)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1194)
	at java.base/java.lang.Thread.run(Thread.java:1474)
Caused by: java.lang.IllegalStateException: Committed
	... 37 more
[...] ERROR org.stripesframework.web.controller.FlashScope - Something is amiss! A timeout occurred while trying to copy a flash scope to a new request. Only StripesFilter should call FlashScope.completeRequest() and FlashScope.beginRequest(), and the calls must be properly synchronized. The timeout likely means that completeRequest() was never called or did not complete successfully on this flash scope.
```

The same code works fine with Jetty 12.0.34.

--- a/jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/internal/DeferredAuthenticationState.java
+++ b/jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/internal/DeferredAuthenticationState.java
@@ -49,22 +49,22 @@ public class DeferredAuthenticationState implements AuthenticationState.Deferred
     {
         try
         {
-            AuthenticationState authenticationState = _authenticator.validateRequest(request, __deferredResponse, Callback.NOT_CALLED);
-            if (authenticationState != null)
+            // Any writes or modifications to the deferred response are ignored.
+            AuthenticationState authenticationState = _authenticator.validateRequest(request, __deferredResponse, Callback.NOOP);
+            if (authenticationState instanceof Succeeded succeeded)
             {
+                // Only cache the AuthenticationState for successful authentication, as no response can be sent.
                 AuthenticationState.setAuthenticationState(request, authenticationState);
-                if (authenticationState instanceof Succeeded succeeded)
-                {
-                    LoginService loginService = _authenticator.getLoginService();
-                    IdentityService identityService = loginService == null ? null : loginService.getIdentityService();
-                    if (identityService != null)
-                    {
-                        UserIdentity user = succeeded.getUserIdentity();
-                        _association = identityService.associate(user, null);
-                    }
 
-                    return succeeded;
+                LoginService loginService = _authenticator.getLoginService();
+                IdentityService identityService = loginService == null ? null : loginService.getIdentityService();
+                if (identityService != null)
+                {
+                    UserIdentity user = succeeded.getUserIdentity();
+                    _association = identityService.associate(user, null);
                 }
+
+                return succeeded;
             }
         }
         catch (ServerAuthException e)

--- a/jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/DeferredAuthenticationTest.java
+++ b/jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/DeferredAuthenticationTest.java
@@ -1,0 +1,171 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.ee10.servlet;
+
+import java.security.Principal;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+import jakarta.servlet.http.HttpServlet;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.security.AuthenticationState;
+import org.eclipse.jetty.security.Authenticator;
+import org.eclipse.jetty.security.Constraint;
+import org.eclipse.jetty.security.HashLoginService;
+import org.eclipse.jetty.security.SecurityHandler;
+import org.eclipse.jetty.security.UserIdentity;
+import org.eclipse.jetty.security.UserStore;
+import org.eclipse.jetty.security.authentication.LoginAuthenticator;
+import org.eclipse.jetty.server.LocalConnector;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.security.Credential;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.jupiter.api.Assertions.assertNull;
+
+public class DeferredAuthenticationTest
+{
+    private Server _server;
+    private LocalConnector _connector;
+    private SecurityHandler.PathMapped _securityHandler;
+    private ServletContextHandler _contextHandler;
+
+    @BeforeEach
+    public void before() throws Exception
+    {
+        _server = new Server();
+        _connector = new LocalConnector(_server);
+        _server.addConnector(_connector);
+
+        _contextHandler = new ServletContextHandler();
+        _server.setHandler(_contextHandler);
+        _contextHandler.setContextPath("/");
+
+        _securityHandler = new SecurityHandler.PathMapped();
+        _securityHandler.put("/private/*", Constraint.ANY_USER);
+        _securityHandler.put("/public/*", Constraint.ALLOWED);
+        _contextHandler.setSecurityHandler(_securityHandler);
+
+        // Configure the loginService with a test user.
+        UserStore userStore = new UserStore();
+        userStore.addUser("test-user", Credential.getCredential("pwd"), new String[]{"user-role"});
+        HashLoginService loginService = new HashLoginService();
+        loginService.setUserStore(userStore);
+        _securityHandler.setLoginService(loginService);
+    }
+
+    public void startServer(Authenticator authenticator, HttpServlet httpServlet) throws Exception
+    {
+        _securityHandler.setAuthenticator(authenticator);
+        _contextHandler.addServlet(httpServlet, "/");
+        _server.start();
+    }
+
+    @AfterEach
+    public void stopServer() throws Exception
+    {
+        _server.stop();
+    }
+
+    @Test
+    public void testWriteOnDeferredAuthentication() throws Exception
+    {
+        AtomicInteger authenticatorCount = new AtomicInteger();
+        AtomicReference<Boolean> authenticatedRef = new AtomicReference<>();
+        AtomicReference<Principal> userPrincipalRef = new AtomicReference<>();
+        AtomicReference<Throwable> servletErrorRef = new AtomicReference<>();
+        startServer(new LoginAuthenticator()
+        {
+            @Override
+            public String getAuthenticationType()
+            {
+                return "TEST";
+            }
+
+            @Override
+            public AuthenticationState validateRequest(Request request, Response response, Callback callback)
+            {
+                authenticatorCount.incrementAndGet();
+                if (!"authenticate".equals(request.getHttpURI().getQuery()))
+                {
+                    // Send 401 response as a challenge.
+                    response.setStatus(HttpStatus.UNAUTHORIZED_401);
+                    response.write(true, BufferUtil.toBuffer("this is a challenge"), callback);
+                    return AuthenticationState.CHALLENGE;
+                }
+
+                if (AuthenticationState.Deferred.isDeferred(response))
+                {
+                    // If we are deferred we attempt to write, this should be ignored.
+                    Response.writeError(request, response, callback, HttpStatus.INTERNAL_SERVER_ERROR_500);
+                    return AuthenticationState.SEND_FAILURE;
+                }
+
+                // Successful authentication.
+                UserIdentity userIdentity = getLoginService().login("test-user", Credential.getCredential("pwd"), request, null);
+                return new UserAuthenticationSucceeded(getAuthenticationType(), userIdentity);
+            }
+        }, new HttpServlet()
+        {
+            @Override
+            protected void service(HttpServletRequest req, HttpServletResponse resp)
+            {
+                try
+                {
+                    assertNull(req.getUserPrincipal());
+                    boolean authenticated = req.authenticate(resp);
+                    authenticatedRef.set(authenticated);
+                    userPrincipalRef.set(req.getUserPrincipal());
+                    if (authenticated)
+                        resp.getWriter().println("success");
+                }
+                catch (Throwable t)
+                {
+                    servletErrorRef.set(t);
+                }
+            }
+        });
+
+        // Authenticator is invoked twice, first time for deferred auth on getUserPrincipal() which tries to write and fails,
+        // the second time for authenticate() which returns false and sends a 401 response.
+        String response = _connector.getResponse("GET /public/foo HTTP/1.0\r\n\r\n");
+        assertThat(response, containsString("401 Unauthorized"));
+        assertThat(response, containsString("this is a challenge"));
+        assertThat(authenticatorCount.get(), equalTo(2));
+        assertThat(authenticatedRef.get(), equalTo(false));
+        assertThat(userPrincipalRef.get(), equalTo(null));
+        assertThat(servletErrorRef.get(), equalTo(null));
+
+        // Authenticator is invoked twice, first time for deferred auth on getUserPrincipal() which tries to write and fails,
+        // the second time for authenticate() which returns true and sends a 200 response.
+        response = _connector.getResponse("GET /public/foo?authenticate HTTP/1.0\r\n\r\n");
+        assertThat(response, containsString("200 OK"));
+        assertThat(response, containsString("success"));
+        assertThat(authenticatorCount.get(), equalTo(4));
+        assertThat(authenticatedRef.get(), equalTo(true));
+        assertThat(userPrincipalRef.get().getName(), equalTo("test-user"));
+        assertThat(servletErrorRef.get(), equalTo(null));
+    }
+}

